src/pkg/signal.go

```diff
@@ -0,0 +1,47 @@
+package pkg
+
+import (
+	"os"
+	"os/signal"
+	"syscall"
+
+	"github.com/rs/zerolog/log"
+)
+
+var (
+	onlyOneSignalHandler = make(chan struct{})
+	stopChannel          = make(chan struct{})
+)
+
+// InitSignalHandler
+// Usage:
+//
+//		func Start() {
+//	   log.Info().Msg("Starting...")
+//	   <-opslevel_common.InitSignalHandler() // Block until signals
+//	   log.Info().Msg("Stopping...")
+//		}
+func InitSignalHandler() <-chan struct{} {
+	close(onlyOneSignalHandler) // panics when called twice
+
+	closeChannel := make(chan os.Signal, 2)
+	signal.Notify(closeChannel, syscall.SIGINT, syscall.SIGTERM)
+	go func() {
+		<-closeChannel
+		close(stopChannel)
+		<-closeChannel
+		os.Exit(1) // second signal. Exit directly.
+	}()
+
+	return stopChannel
+}
```

Comment:
we've now duplicated this code 2 times.  If we do this a 3rd time (which is highly likely given the expected architecture of opslevel-agent) the work you've done to remove opslevel-common was probably not a great idea.

I now questioning if we should have put this in https://github.com/OpsLevel/opslevel-k8s-controller as a `pkg/signal` since both this and kubectl-opslevel (and likely opslevel-agent) will all need.

Reply:
This code is not used in `kubectl-opslevel`, the signal handler function that is used [there is a different function](https://github.com/OpsLevel/kubectl-opslevel/blob/5dd7c31db0973b5466c822147aa832892882495a/src/common/signal.go#L13) that returns a `Context`  instead of a `<- chan struct{}`.

A copy of the `signal.go` that was ripped out of opslevel-common [lives in that repo under the pkg/ folder](https://github.com/OpsLevel/kubectl-opslevel/blob/5dd7c31db0973b5466c822147aa832892882495a/src/pkg/signal/signal.go) solely for posterity in case we needed it*. I just copied the file here into this repo to be able to stop compilation errors after removing references to opslevel-common because changing signal handling logic is out of scope for this PR. 

It is very likely that we will not end up using this file altogether once this updated version opslevel-runner has been properly tested and prepared for the next release given that we have existing `context.TODO()` calls that need to be taken care of and there are new places where contexts are now required because of deprecations, like in k8s.go.

It's very likely that every program we write will have slightly different signal handling needs and so it is not practical to introduce a dependency for signal handling. The signal handling code provided by this file is only 14 lines long - the `Run()` and `Stop()` functions were not being called at all in this repo and [in kubectl-opslevel they were not working.](https://github.com/OpsLevel/kubectl-opslevel/pull/235#discussion_r1453757778)

I don't think it would be necessary for the K8s controller repo to provide a signal handling function because its core responsibility is to monitor for relevant events whether the program calling it needs to handle signals or not. 

*I would actually prefer we just removed this file altogether both here and in kubectl-opslevel.

Reply:
On this `InitSignalHandler` and the [kubectl-opslevel InitSignalHandler](https://github.com/OpsLevel/kubectl-opslevel/blob/5dd7c31db0973b5466c822147aa832892882495a/src/common/signal.go#L13-L24) - it is confusing if nothing else.

They are named the same which tells the reader that they are closely related. While they have different function signatures and return different types, they accomplish a very similar task. Clearly we need a signal handler in different tools and those tools likely have unique, albeit related needs.

If it is likely that these functions will be removed we still need something for now. At the moment, it seems to make a lot of sense to consolidate `InitSignalHandler` functions into a common package like `pkg/signal`. And in a common repo that rarely requires moving code from repo A to repo B (and repo C maybe).

Reply:
You are assuming that they are suppose to be different but they are not.  They both should be handling signal handling the exact same way because they both a run as processes inside a container in a k8s pod.  So they have to respond to k8s pod signals the exact same way.

Reply:
Late update - after reading this discussion back in Jan and following up afterwards in a call, I created: https://github.com/OpsLevel/team-platform/issues/212. That ticket also links to a slack convo where we discussed this. 

I have removed all destructive changes from this PR so am requesting re-review, will resolve this conversation unless someone strongly objects.